### PR TITLE
Fix health check for gRPC

### LINKED ISSUE #1662: ALB GRPC target not working: potentially incorrect healthCheckConfig.matcher in model generated for gRPC target groups
In #1589, gRPC support has been added to this controller. While this is not yet in a release, we are trying this out for an ongoing project, and hence we have built an image ourselves off [`e9691e231f2305fc91de1838d2de5386d9dfae53`](https://github.com/kubernetes-sigs/aws-load-balancer-controller/tree/e9691e231f2305fc91de1838d2de5386d9dfae53).

After installation (with image modified in the [release manifest](https://raw.githubusercontent.com/kubernetes-sigs/aws-alb-ingress-controller/main/docs/install/v2_0_0_full.yaml)), the ingress in our EKS cluster has been configured as followed:

```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: REDACTED
  namespace: default
  annotations:
    kubernetes.io/ingress.class: alb
    alb.ingress.kubernetes.io/group.name: REDACTED
    alb.ingress.kubernetes.io/tags: ingress-app=REDACTED
    alb.ingress.kubernetes.io/target-type: ip
    alb.ingress.kubernetes.io/scheme: internal
    alb.ingress.kubernetes.io/load-balancer-attributes: access_logs.s3.enabled=true,access_logs.s3.bucket=REDACTED,access_logs.s3.prefix=REDACTED,routing.http2.enabled=true
    alb.ingress.kubernetes.io/listen-ports: '[{"HTTPS": 443}]'
    alb.ingress.kubernetes.io/ip-address-type: ipv4
    alb.ingress.kubernetes.io/inbound-cidrs: '10.0.0.0/8'
    alb.ingress.kubernetes.io/ssl-policy: ELBSecurityPolicy-TLS-1-2-2017-01
    alb.ingress.kubernetes.io/backend-protocol: HTTP
    alb.ingress.kubernetes.io/backend-protocol-version: GRPC
    alb.ingress.kubernetes.io/healthcheck-path: '/grpc.health.v1.Health/Check'
    alb.ingress.kubernetes.io/healthcheck-port: '8080'
    alb.ingress.kubernetes.io/success-codes: '0,5,12,16'
  labels:
    app: REDACTED
spec:
  rules:
    - host: REDACTED
      http:
        paths:
          - path: /*
            backend:
              serviceName: REDACTED
              servicePort: 8080
```

This is backed by a service as followed:

```
apiVersion: v1
kind: Service
metadata:
  name: REDACTED
  namespace: default
spec:
  clusterIP: 172.20.73.46
  ports:
  - port: 8080
    protocol: TCP
    targetPort: 8080
  selector:
    app: REDACTED
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}
```

Which, in turn, proxies to `go-grpc` backend pods.

However, after setting this up (and killing the load balancer controller pod several times to be sure), we observed the following in ingress status:

```
  Warning  FailedDeployModel  19m  ingress  Failed deploy model due to ValidationError: Health check matcher GRPC code cannot be empty
           status code: 400, request id: b0be8cb8-e37e-42e9-b2fe-4604ed8f9b68
  Warning  FailedDeployModel  80s (x10 over 19m)  ingress  (combined from similar events): Failed deploy model due to ValidationError: Health check matcher GRPC code cannot be empty
           status code: 400, request id: cbca6b5f-84ee-4504-ae55-6164f2592357
```

And in the controller pod logs:

```
{"level":"info","ts":1605637136.7691634,"logger":"controllers.ingress","msg":"successfully built model","model":"SEE BELOW"}
{"level":"info","ts":1605637136.9938061,"logger":"controllers.ingress","msg":"creating targetGroup","stackID":"REDACTED","resourceID":"default/REDACTED-REDACTED:8080"}
{"level":"error","ts":1605637137.1319094,"logger":"controller","msg":"Reconciler error","controller":"ingress","name":"REDACTED","namespace":"","error":"ValidationError: Health check matcher GRPC code cannot be empty\n\tstatus code: 400, request id: cbca6b5f-84ee-4504-ae55-6164f2592357"}
```

I expanded the model output, and I believe the followed is the relevant bit:

```
        (...)
        },
        "AWS::ElasticLoadBalancingV2::TargetGroup": {
            "default/REDACTED-REDACTED:8080": {
                "spec": {
                    "name": "k8s-default-k8smonit-7de211cf8a",
                    "targetType": "ip",
                    "port": 8080,
                    "protocol": "HTTP",
                    "protocolVersion": "GRPC",
                    "healthCheckConfig": {
                        "port": 8080,
                        "protocol": "HTTP",
                        "path": "/grpc.health.v1.Health/Check",
                        "matcher": {
                            "httpCode": "0,5,12,16"
                        },
                        "intervalSeconds": 15,
                        "timeoutSeconds": 5,
                        "healthyThresholdCount": 2,
                        "unhealthyThresholdCount": 2
                    },
                    "tags": {
                        "ingress-app": "REDACTED"
                    }
                }
            }
        },
(...)
```

On consulting AWS documentation, it looks like we should send `grpcCode` rather than `httpCode`:
https://docs.aws.amazon.com/elasticloadbalancing/latest/APIReference/API_Matcher.html

Briefly consulting the code, while I don't have time to dig in, it looks like #1589 might have missed handling the GRPC case for [`HealthCheckMatcher`](https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/e9691e231f2305fc91de1838d2de5386d9dfae53/pkg/model/elbv2/target_group.go#L63) via [`buildTargetGroupHealthCheckMatcher`](https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/e9691e231f2305fc91de1838d2de5386d9dfae53/pkg/ingress/model_build_target_group.go#L235).

Please let me know if you need any further information.

### PR DESCRIPTION
**Pull request by backjo**
Fixes #1662

### CODE REVIEW

File: pkg/deploy/elbv2/target_group_manager.go [hunk 1]
```diff
@@ -233,7 +233,12 @@ func buildSDKModifyTargetGroupInput(tgSpec elbv2model.TargetGroupSpec) *elbv2sdk
 	return sdkObj
 }
 
-func buildSDKMatcher(modelMatcher elbv2model.HealthCheckMatcher) *elbv2sdk.Matcher {
+func buildSDKMatcher(modelMatcher elbv2model.HealthCheckMatcher, protocolVersion *elbv2model.ProtocolVersion) *elbv2sdk.Matcher {
```
Comment (M00nF1sh) [inline pkg/deploy/elbv2/target_group_manager.go:236]: i think this is incorrect.
we should use modelMatcher, and check whether `modelMatcher.GPRC` and `modelMatcher.HTTPCode` is empty or not. (ideally we should change GRPC/HTTPCode to a string pointer as well, but keep it as string works as well)

Comment (backjo) [reply pkg/deploy/elbv2/target_group_manager.go:236]: yeah, I was going between that option and this one myself. Makes sense.

**Comment by backjo** [comment]
Addressed the review comment - and also fixed failing unit tests. Apparently, unit tests are not being run on PRs anymore (issue opened on #1665 )

File: pkg/model/elbv2/target_group.go [hunk 2]
```diff
@@ -60,7 +60,10 @@ const (
 // Information to use when checking for a successful response from a target.
 type HealthCheckMatcher struct {
 	// The HTTP codes.
-	HTTPCode string `json:"httpCode"`
+	HTTPCode *string `json:"httpCode,omitempty"`
```
Comment (M00nF1sh) [inline pkg/model/elbv2/target_group.go:63]: nice :D

File: pkg/deploy/elbv2/target_group_manager.go [hunk 3]
```diff
@@ -165,7 +165,9 @@ func isSDKTargetGroupHealthCheckDrifted(tgSpec elbv2model.TargetGroupSpec, sdkTG
 	if hcConfig.Path != nil && awssdk.StringValue(hcConfig.Path) != awssdk.StringValue(sdkObj.HealthCheckPath) {
 		return true
 	}
-	if hcConfig.Matcher != nil && (sdkObj.Matcher == nil || hcConfig.Matcher.HTTPCode != awssdk.StringValue(sdkObj.Matcher.HttpCode)) {
+	if (hcConfig.Matcher != nil && sdkObj.Matcher == nil) ||
```
Comment (M00nF1sh) [inline pkg/deploy/elbv2/target_group_manager.go:168]: this logic should be
```
if (hcConfig.Matcher != nil && (sdkObj.Matcher == nil ||  awssdk.String(hcConfig.Matcher.GRPCCode) != awssdk.StringValue(sdkObj.Matcher.GrpcCode) || awssdk.String(hcConfig.Matcher.HTTPCode) != awssdk.StringValue(sdkObj.Matcher.HttpCode))
```

Comment (backjo) [reply pkg/deploy/elbv2/target_group_manager.go:168]: ah good catch

**Review by M00nF1sh — APPROVED** [review]
/lgtm 
👍

**Pull request merged**